### PR TITLE
Migrate to Swift 3

### PR DESCRIPTION
**Pull request by superpeteblaze**
## What changed
- Update directions target SWIFT_VERSION and test targets 
- Run Swift 3 migration on directions framework
- Convert example to swift 3
## Todos
- [x] Integrate with a Polyline swift3 branch
- [x] Swiftify the API
- [x] Document
## Notify

@1ec5 
## Note
- This of course won't build without a swift3 version of Polyline

### CODE REVIEW

**Comment by superpeteblaze** [comment]
@1ec5, I have forked Polyline (no access to source), and have a swift 3 branch up which seems to work as expected with MapboxDirections

File: MapboxDirections/MBRouteOptions.swift [hunk 1]
```diff
@@ -33,21 +33,21 @@ public enum RouteShapeFormat: UInt, CustomStringConvertible {
      
      This standard format is human-readable and can be parsed straightforwardly, but it is far more verbose than `Polyline`.
```
Comment (1ec5) [inline MapboxDirections/MBRouteOptions.swift]: Lowercase `polyline`.

File: MapboxDirections/MBRouteOptions.swift [hunk 2]
```diff
@@ -33,21 +33,21 @@ public enum RouteShapeFormat: UInt, CustomStringConvertible {
...
      */
-    case GeoJSON
+    case geoJSON
     /**
      The route’s shape is delivered in [encoded polyline algorithm](https://developers.google.com/maps/documentation/utilities/polylinealgorithm) format.
      
      This machine-readable format is considerably more compact than `GeoJSON`.
```
Comment (1ec5) [inline MapboxDirections/MBRouteOptions.swift]: Lowercase `geoJSON`.

File: MapboxDirections/MBRouteOptions.swift [hunk 3]
```diff
@@ -74,29 +74,29 @@ public enum RouteShapeResolution: UInt, CustomStringConvertible {
...
      The route’s shape is simplified.
      
      This resolution considerably reduces the size of the response. The resulting shape is suitable for display at a low zoom level, but it lacks the detail necessary for focusing on individual segments of the route.
      */
-    case Low
+    case low
     /**
      The route’s shape is as detailed as possible.
      
      The resulting shape is equivalent to concatenating the shapes of all the route’s consitituent steps. You can focus on individual segments of this route while faithfully representing the path of the route. If you only intend to show a route overview and do not need to analyze the route line in any way, consider specifying `Low` instead to considerably reduce the size of the response.
```
Comment (1ec5) [inline MapboxDirections/MBRouteOptions.swift]: Lowercase `low`.

File: MapboxDirections/MBRouteOptions.swift [hunk 4]
```diff
@@ -225,7 +225,7 @@ public class RouteOptions: NSObject {
      
      The default value of this property is `Polyline`.
```
Comment (1ec5) [inline MapboxDirections/MBRouteOptions.swift]: Lowercase `polyline`.

File: MapboxDirections/MBRouteOptions.swift [hunk 5]
```diff
@@ -234,7 +234,7 @@ public class RouteOptions: NSObject {
      
      The default value of this property is `Low`, specifying a low-resolution route shape.
```
Comment (1ec5) [inline MapboxDirections/MBRouteOptions.swift]: Lowercase `low`.

File: MapboxDirections/MBRouteOptions.swift [hunk 6]
```diff
@@ -358,7 +358,7 @@ public class RouteOptionsV4: RouteOptions {
      
      By default, the value of this property is `Text`, specifying plain text instructions.
```
Comment (1ec5) [inline MapboxDirections/MBRouteOptions.swift]: Lowercase `text`.

File: MapboxDirections/MBRouteStep.swift [hunk 7]
```diff
@@ -115,73 +115,73 @@ public enum ManeuverType: Int, CustomStringConvertible {
...
      
      The maneuver direction indicates which fork to take.
      */
-    case ReachFork
+    case reachFork
     
     /**
      The step requires the user to turn at either a T-shaped three-way intersection or a sharp bend in the road where the road also changes names.
      
      This maneuver type is called out separately so that the user may be able to proceed more confidently, without fear of having overshot the turn. If this distinction is unimportant to you, you may treat the maneuver as an ordinary `Turn`.
```
Comment (1ec5) [inline MapboxDirections/MBRouteStep.swift]: Lowercase `turn`.

File: MapboxDirections/MBRouteStep.swift [hunk 8]
```diff
@@ -115,73 +115,73 @@ public enum ManeuverType: Int, CustomStringConvertible {
...
      
      The exit index indicates the number of roundabout exits up to and including the exit that the user must take.
      */
-    case TakeRoundabout
+    case takeRoundabout
     
     /**
      The step requires the user to enter and exit a roundabout (traffic circle or rotary) that is compact enough to constitute a single intersection.
      
      This maneuver type is called out separately because the user may perceive the roundabout as an ordinary intersection with an island in the middle. If this distinction is unimportant to you, you may treat the maneuver as either an ordinary `Turn` or as a `TakeRoundabout`.
```
Comment (1ec5) [inline MapboxDirections/MBRouteStep.swift]: Lowercase `turn` and `takeRoundabout`.

File: MapboxDirections/MBRouteStep.swift [hunk 9]
```diff
@@ -190,53 +190,53 @@ public enum ManeuverType: Int, CustomStringConvertible {
      
      Similar changes can occur simultaneously with other maneuvers, such as when the road changes its name at the site of a movable bridge. In such cases, `HeedWarning` is suppressed in favor of another maneuver type.
```
Comment (1ec5) [inline MapboxDirections/MBRouteStep.swift]: Lowercase `heedWarning`.

**Comment by 1ec5** [comment]
Note that once raphaelmor/Polyline upgrades to Swift 3, you’ll need to update the podspec to require a newer version of that library.

**Comment by superpeteblaze** [comment]
Swift 3 has changed to using lower case enums - see here. Also, I asked for write access to Polyline so received no response. So I forked Polyline and pointed the podspec at the migrated fork.

**Comment by 1ec5** [comment]
The podspec only covers the example application in this repository. The podfile still points to the official repository, and to my knowledge, there’s no way to automatically point it elsewhere. Every developer integrating this library will need to add your Polyline fork as a dependency ahead of any pod that uses Polyline. We’ll need a blurb in the readme to that effect. (For example, this library [used to depend](https://github.com/mapbox/MapboxDirections.swift/blob/97cfb5eb4d226c29ede2b5ccef27f920b1ba5696/README.md#installation) on my fork of RequestKit.)

**Comment by superpeteblaze** [comment]
@1ec5 Apologies, you are correct. I did the same with my app. I added my Polyline fork as a dependency ahead of the original Polyline. I'll add the blurb to the readme

**Comment by superpeteblaze** [comment]
@1ec5 I updated the readme to include the forked Polyline. However, I am unsure about how to handle Carthage. Any ideas?

**Comment by 1ec5** [comment]
I _think_ you’d [add your branch](https://github.com/Carthage/Carthage/blob/81880f60e63b58d9dfaf0e489473e27dfd670d50/Documentation/Artifacts.md#version-requirement) in place of the existing Polyline entry:

```cartfile
github "https://github.com/superpeteblaze/Polyline" "swift3"
```

**Comment by superpeteblaze** [comment]
@1ec5 Updated with the latest Beta 6 Swift language changes

File: MapboxDirections/MBDirections.swift [hunk 10]
```diff
@@ -183,13 +183,14 @@ public class Directions: NSObject {
...
-    private func dataTaskWithURL(_ url: URL, completionHandler: (json: JSONDictionary) -> Void, errorHandler: (error: NSError) -> Void) -> URLSessionDataTask {
+    private func dataTaskWithURL(_ url: URL, completionHandler: @escaping (_ json: JSONDictionary) -> Void, errorHandler: @escaping (_ error: NSError) -> Void) -> URLSessionDataTask {
+        
         let request = NSMutableURLRequest(url: url)
         request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
         return URLSession.shared.dataTask(with: request as URLRequest) { (data, response, error) in
             var json: JSONDictionary = [:]
             if let data = data {
-                do {
+                do  {
```
Comment (1ec5) [inline MapboxDirections/MBDirections.swift]: Nit: an extra space snuck in.

**Comment by superpeteblaze** [comment]
@1ec5 Ran Xcode GM Swift migrator

**Comment by Eldorado234** [comment]
I have no idea why one of the tvOS builds now fails:

> ❌  /Users/vagrant/git/MapboxDirections/MBRouteLeg.swift:1:8: no such module 'Polyline'
> import Polyline

**Comment by 1ec5** [comment]
Ugh, that’s the same error that’s been failing the watchOS build bot.

**Comment by 1ec5** [comment]
221e451083961894b9ef5db58fc762e1aa295a7f fixes the tvOS and watchOS builds by replacing the superpeteblaze/Polyline dependency with raphaelmor/Polyline@f86f34dac9aae7e027fd9f62c2e01dd77f1ee312, which includes raphaelmor/Polyline#38. (Polyline’s master branch is currently broken: raphaelmor/Polyline#40.)

**Comment by 1ec5** [comment]
This PR’s swift3 branch was so long-lived that, out of expediency, we suggested that developers pin their Podfile or Cartfile directly to the branch (as opposed to a commit on the branch). Now that the swift3 branch has been merged to master, 5384a42d8d18ff8e2e4520064f0bbc8da560c89d logs a warning to the console urging the developer to move to the master branch. The swift3 branch is no longer being maintained and will be deleted soon. If you pinned to a specific commit on this branch, that commit will continue to exist because it was merged into master. Otherwise, we urge you to heed the warning and move to master, and thence to the upcoming release once it’s published.

**Pull request merged**